False positive when accessing a TypedDict in a list comprehension

**Bug Report**

When using a TypedDict in list comprehension 

**To Reproduce**

```python
from typing import TypedDict

class Data (TypedDict):
    field1:int
    field2:str
    field3:str

data:Data = {
    'field1':1,
    'field2':'string',
    'field3':'string'
}

stringValues = [value for value in ("field1", "field2", "field3") if type(data[value]) is str] # error here
```

**Expected Behavior**

Works as the non list comprehension verson:
```python
from typing import TypedDict

class Data (TypedDict):
    field1:int
    field2:str
    field3:str

data:Data = {
    'field1':1,
    'field2':'string',
    'field3':'string'
}
for value in ("field1", "field2", "field3"):
    if type(data[value]) is str:
        print(f"{value} is a string")
```

**Actual Behavior**

mypy complains about the typing of value in the list comprehension. Error reported is
```
error: TypedDict key must be a string literal; expected one of ("field1", "field2", "field3")  [literal-required]
```

**Your Environment**

<!-- Include as many relevant details about the environment you experienced the bug in -->

- Mypy version used: mypy 1.16.0 (compiled: yes)
- Mypy command-line flags: NA
- Mypy configuration options from `mypy.ini` (and other config files): NA
- Python version used: 3.13.5

--- a/mypy/semanal.py
+++ b/mypy/semanal.py
@@ -6432,7 +6432,7 @@ class SemanticAnalyzer(
             if i > 0:
                 sequence.accept(self)
             # Bind index variables.
-            self.analyze_lvalue(index)
+            self.analyze_lvalue(index, is_index_var=True)
             for cond in conditions:
                 cond.accept(self)
 

--- a/test-data/unit/check-typeddict.test
+++ b/test-data/unit/check-typeddict.test
@@ -779,6 +779,17 @@ def get_coordinate(p: TaggedPoint, key: str) -> Union[str, int]:
 [builtins fixtures/dict.pyi]
 [typing fixtures/typing-typeddict.pyi]
 
+[case testCanGetItemOfTypedDictWithStringLiteralKeyInComprehension]
+from typing import TypedDict
+Data = TypedDict('Data', {'field1': int, 'field2': str, 'field3': str})
+data: Data
+def is_str(value: object) -> bool: ...
+filtered_keys1 = [data[key] for key in ("field1", "field2", "field3")]
+filtered_keys2 = [key for key in ("field1", "field2", "field3") if is_str(data[key])]
+reveal_type(filtered_keys1)    # N: Revealed type is "builtins.list[builtins.int | builtins.str]"
+reveal_type(filtered_keys2)    # N: Revealed type is "builtins.list[builtins.str]"
+[builtins fixtures/for.pyi]
+[typing fixtures/typing-full.pyi]
 
 -- Special Method: __setitem__
 

